File: Python/generate_controller.py
#! /usr/bin/python

import sys

print('controller generator')

def indent(i,str):
    space=' '
    return (space * i)+str

def guess_singular(plural):
    guessed=plural[0:-1]
    print(  "\nsuggested singular for plural %s is: %s" % (plural, guessed))
    print( "If it is correct press Enter,")
    print( "but if it not correct please enter correct version ")
    entered = raw_input('')
    if entered == '':
        return guessed
    else:
        return entered

def to_underscores(camel_string):
    cap_str='ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    capitals = []
    result = ''
    for x in cap_str:
        capitals.append(x)
    for letter in camel_string:
        if letter.isupper():
            result += ('_' + letter.lower())
        else:
            result += letter
    return result.strip('_')

def spaces(n):
    return ' ' * n   

def generate(sing_camel, plu_camel):
    sing_under = to_underscores(sing_camel)
    plu_under =  to_underscores(plu_camel)
    methods = [
        ['index','all', False, False],             
        ['show','find',':id', False],              
        ['new','new', False, False],
        ['edit','find',':id', False],
        ['create','new',":{0}".format(sing_under),"if {0}.save".format(sing_under)],              
        ['update','find',':id',"if @{0}.update_attributes(params[:{0}])".format(sing_under)],
        ['destroy','find',':id', False] ]
    print methods
    r = "class {0}Controller < ApplicationController\n".format(plu_camel)
    r += indent(2, "respond_to :html, :json, :xml\n")
    for e in methods:
        r += indent(2, "def {0}\n".format(e[0]))
        if e[0] == 'index':
            obj = plu_under
        else:
            obj = sing_under
        r += indent(4, "@{0} = {1}.{2}".format(obj, sing_camel, e[1]))
        if e[2]:
            r += "(params[{0}])\n".format(e[2])
        else:
            r += "\n"
        if e[3]:
            r += indent(4,"{0}\n{2}\n{1}else\n{2}\n{1}end\n".format(e[3], spaces(4), spaces(6)))
        if e[0] == 'destroy':
            r += indent(4, "@{}.destroy\n".format(obj))
        r += indent(4, "respond_with(@{0})\n".format(obj))
        r += indent(2, "end\n")
    r += "end\n"
    print(r)

def main():
    print 'in main'
    plural_name = 'StruggleReports'
    singular_name = guess_singular(plural_name)
    generate( singular_name, plural_name)

main()

File: Ruby/generate_controller.rb
#! /usr/bin/env ruby

class String
  def to_underscores
    # converts CamelCase to undescores
    capitals = %w(A B C D E F G H I J K L M N O P Q R S T U V W X Y Z)
    result=''
    self.size.times do |x|
      letter = self[x,1]
      if capitals.index letter
        result += "_"+letter.downcase
      else
        result += letter
      end
    end
    result[0,1]='' if result[0,1]=='_'
    return result
  end
end
class Fixnum
  def spaces; ' ' * self; end
end

def guess_singular(plural)
  guessed=plural[0..-2]
  puts "\nsuggested singular for #{plural} is: "+guessed
  puts "If it is correct press Enter,"
  puts "but if it not correct please enter correct version"
  entered=STDIN.gets.chomp
  puts
  return guessed  if entered == ''
  return entered
end
def indent (i,str)
  i.spaces+str
end
def generate(plu_camel,plu_under,sing_camel,sing_under)
  methods=[ ['index','all'],
            ['show','find',':id'],
            ['new','new'],
            ['edit','find',':id'],
            ['create','new',":#{sing_under}","if @#{sing_under}.save"],
            ['update','find',':id',"if @#{sing_under}.update_attributes(params[:#{sing_under}]) "],
            ['destroy','find',':id'] ]

  r="class #{plu_camel}Controller < ApplicationController\n"
  r << indent( 2, "respond_to :html, :json, :xml\n")
  methods.each do |e|
    r << indent( 2,"def #{e[0]}\n")
    e[0]=='index' ? obj=plu_under : obj=sing_under
    r << indent( 4,"@#{obj} = #{sing_camel}.#{e[1]}")
    e[2] ? r << "(params[#{e[2]}])\n" : r << "\n"
    r << indent( 4,"#{e[3]}\n#{6.spaces}#\n#{4.spaces}else\n#{6.spaces}#\n#{4.spaces}end\n") if e[3]
    r << indent( 4,"@#{obj}.destroy\n") if e[0]=='destroy'
    r << indent( 4,"respond_with(@#{obj})\n")
    r << indent( 2,"end\n")
  end
  r << "end\n"
  puts r
end

def usage
  puts "usage:\n #{$0} PluralControllerNames"
end
#################################################
#puts $*.inspect
plural_name = $*[0]
if plural_name==nil or plural_name[0,1] == '-'
  usage
elsif plural_name != nil
  singular_name = guess_singular( plural_name)
  generate(plural_name,plural_name.to_underscores,singular_name,singular_name.to_underscores)
else
  usage
end

